### PR TITLE
Fix/ddp val barrier

### PR DESCRIPTION
**Pull request by tedo001**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
This PR fixes distributed validation synchronization and improves segmentation result handling by safely propagating optional semantic segmentation outputs, while also hardening a pose augmentation edge case. 🛠️

### 📊 Key Changes
- Added a `dist.barrier()` before validation in DDP training to synchronize ranks and avoid NCCL collective mismatches during validation.
- Set `broadcast_buffers=False` in `DistributedDataParallel` setup to reduce unwanted buffer synchronization behavior during multi-GPU training.
- Extended `Results` to store optional `semseg` tensors for auxiliary semantic segmentation logits.
- Updated YOLO segmentation inference to extract, carry, and return optional semantic segmentation logits alongside boxes and masks.
- Adjusted segmentation head and proto block logic so optional semseg outputs are returned when available, not only during training.
- Hardened `v8_transforms()` keypoint flip validation by guarding against `kpt_shape=None` before checking `flip_idx` length.

### 🎯 Purpose & Impact
- Improves DDP training stability, especially around validation, by preventing rank desynchronization issues. ⚡
- Enables downstream access to auxiliary semantic segmentation logits in inference results, which can help advanced segmentation workflows.
- Preserves export behavior while expanding non-export inference outputs more safely.
- Prevents a potential error path in pose/keypoint augmentation configs when `flip_idx` is provided without a defined `kpt_shape`. ✅

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @tedo001, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated review note to help streamline triage, and an engineer will assist you soon. Please review the checklist below:

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀🙂

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -517,6 +519,8 @@ def _do_train(self):
             # Validation
             final_epoch = epoch + 1 >= self.epochs
             if self.args.val or final_epoch or self.stopper.possible_stop or self.stop:
+                if RANK != -1:
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py:522]: ⚠️ **HIGH**: This barrier will deadlock if it's executed from the existing rank-0 validation path in `_do_train`: all DDP ranks must enter the same `dist.barrier()`, but nonzero ranks do not validate here. As written, rank 0 will block forever the first time validation is triggered. Move the synchronization point outside the rank-gated validation block, or make every rank participate in the barrier.

File: ultralytics/engine/results.py [hunk 2]
```diff
@@ -259,6 +260,7 @@ def __init__(
         self.probs = Probs(probs) if probs is not None else None
         self.keypoints = Keypoints(keypoints, self.orig_shape) if keypoints is not None else None
         self.obb = OBB(obb, self.orig_shape) if obb is not None else None
+        self.semseg = semseg  # auxiliary semantic segmentation logits (C, H, W) or None
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/results.py:263]: 💡 **MEDIUM**: `Results` now stores `semseg`, but this attribute is not integrated into the existing tensor-handling machinery. Methods like `cpu()`, `cuda()`, `to()`, and `numpy()` typically rely on the tracked result keys, so `semseg` will be left behind on its original device/type while the rest of the result is converted. That can leak GPU memory and produce inconsistent `Results` objects after conversion.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Good targeted fixes overall, but two issues need attention: the new DDP barrier is very likely to hang validation unless every rank reaches it, and `Results.semseg` needs to be wired into the existing result-conversion path so it behaves like the other tensor attributes.

💬 Posted 2 inline comments
